### PR TITLE
Make Atom an abstract class

### PR DESCRIPTION
**Pull request by madmike200590**
`Atom` changed from an interface to abstract class - This allows us to make `equals` and `hashCode` abstract in `Atom`, thereby enforcing these to be implemented by all implementations of `Atom`

Also contains implementations for dependency and component graphs, same as currently open PR #185  (ideally merge this one before #185 )

### CODE REVIEW

File: src/main/java/at/ac/tuwien/kr/alpha/Alpha.java [hunk 1]
```diff
@@ -71,11 +80,36 @@ public Program readProgram(InputConfig cfg) throws IOException {
...
+			// FIXME ProgramAnalysis is again created in grounder, should generally do this
+			// in Alpha and pass into grounder (not too critical, only done here in case
+			// user requests dot output for dependency graph)
+			ProgramAnalysis analysis = new ProgramAnalysis(retVal);
+			new DependencyGraphWriter().writeAsDot(analysis.getDependencyGraph(), cfg.getDepGraphTarget(), this.config.isDebugInternalChecks());
 		}
 		return retVal;
 	}
 
+	private Program doProgramTransformations(Program program) {
```
Comment (AntoniusW) [inline src/main/java/at/ac/tuwien/kr/alpha/Alpha.java]: I do not fully understand why this code is now duplicated (here and in `NaiveGrounder`)? It seems to be an intermediate version of #185 where program transformations are partially refactored. Can you remove from this PR the partial refactoring of the program transformations?

File: src/main/java/at/ac/tuwien/kr/alpha/common/atoms/Atom.java [hunk 2]
```diff
@@ -132,4 +133,13 @@ default int compareTo(Atom o) {
 		return 0;
 	}
 
+	/**
+	 * Returns whether this Atom is a builtin-atom according to the ASP standard. This covers the atoms currently defined by <code>ComparisonAtom</code>.
+	 * 
+	 * @return true if this Atom is a comparison atom, false otherwise.
```
Comment (AntoniusW) [inline src/main/java/at/ac/tuwien/kr/alpha/common/atoms/Atom.java]: This comment does not really describe the code below (it describes the intent of the method though, but that should rather be part of the description above).

File: src/main/java/at/ac/tuwien/kr/alpha/grounder/NaiveGrounder.java [hunk 3]
```diff
@@ -96,15 +119,18 @@ public NaiveGrounder(Program program, AtomStore atomStore, boolean debugInternal
 		super(filter, bridges);
 		this.atomStore = atomStore;
 
-		programAnalysis = new ProgramAnalysis(program);
-		analyzeUnjustified = new AnalyzeUnjustified(programAnalysis, atomStore, factsFromProgram);
+		this.applyProgramTransformations(program, useCountingGrid);
+		// TODO it might make more sense to do this outside the grounder, but where
```
Comment (AntoniusW) [inline src/main/java/at/ac/tuwien/kr/alpha/grounder/NaiveGrounder.java]: This looks again like a half-refactored intermediate from #185. If possible, remove/revert it please.

File: src/main/resources/logback.xml [hunk 4]
```diff
@@ -13,11 +13,11 @@
             <pattern>%-4r %-5level %logger{35}: %msg%n</pattern>
         </encoder>
     </appender>
-    <root level="INFO">
```
Comment (AntoniusW) [inline src/main/resources/logback.xml]: Please keep the log level at INFO.

File: src/test/resources/partial-eval/components-test.asp [hunk 5]
```diff
@@ -0,0 +1,13 @@
+f0.
```
Comment (AntoniusW) [inline src/test/resources/partial-eval/components-test.asp]: Is this file needed?

File: src/main/java/at/ac/tuwien/kr/alpha/grounder/structure/AnalyzeUnjustified.java [hunk 6]
```diff
@@ -89,8 +91,7 @@ private ReturnExplainUnjust explainUnjust(LitSet x, Assignment currentAssignment
 		// Construct set of all 'rules' such that head unifies with p.
 		List<RuleAndUnifier> rulesUnifyingWithP = rulesHeadUnifyingWith(p);
 		log("Rules unifying with {} are {}", p, rulesUnifyingWithP);
-		rulesLoop:
-		for (RuleAndUnifier ruleUnifier : rulesUnifyingWithP) {
+		rulesLoop: for (RuleAndUnifier ruleUnifier : rulesUnifyingWithP) {
```
Comment (AntoniusW) [inline src/main/java/at/ac/tuwien/kr/alpha/grounder/structure/AnalyzeUnjustified.java]: Please leave the loop labels on their own line.

File: src/main/java/at/ac/tuwien/kr/alpha/grounder/structure/AnalyzeUnjustified.java [hunk 7]
```diff
@@ -197,8 +198,7 @@ private ReturnExplainUnjust explainUnjust(LitSet x, Assignment currentAssignment
 
 			log("Checking atoms over predicate: {}", b.getPredicate());
 			AssignedAtomsIterator assignedAtomsOverPredicate = getAssignedAtomsOverPredicate(b.getPredicate());
-			atomLoop:
```
Comment (AntoniusW) [inline src/main/java/at/ac/tuwien/kr/alpha/grounder/structure/AnalyzeUnjustified.java]: Moved loop label again.

File: src/main/java/at/ac/tuwien/kr/alpha/grounder/structure/AnalyzeUnjustified.java [hunk 8]
```diff
@@ -298,13 +298,13 @@ public Atom next() {
...
 	}
 
-
 	private List<RuleAndUnifier> rulesHeadUnifyingWith(Atom p) {
 
 		List<RuleAndUnifier> rulesWithUnifier = new ArrayList<>();
 		Predicate predicate = p.getPredicate();
 		// Check if literal is built-in with a fixed interpretation.
-		if (p instanceof FixedInterpretationLiteral) {
+		// if (p instanceof FixedInterpretationLiteral) {
```
Comment (AntoniusW) [inline src/main/java/at/ac/tuwien/kr/alpha/grounder/structure/AnalyzeUnjustified.java]: This looks dangerous. I do not think it is correct. What is the rationale for the change?

File: src/main/java/at/ac/tuwien/kr/alpha/grounder/structure/ProgramAnalysis.java [hunk 9]
```diff
@@ -1,31 +1,81 @@
...
+		this.initializeRules(program);
+		this.dependencyGraph = DependencyGraph.buildDependencyGraph(this.nonGroundRules);
+	}
+
+	private void initializeFacts(Program program) {
+		List<Instance> tmpInstances;
+		Predicate tmpPredicate;
+		for (Atom fact : program.getFacts()) {
+			tmpInstances = FactIntervalEvaluator.constructFactInstances(fact);
+			tmpPredicate = fact.getPredicate();
```
Comment (AntoniusW) [inline src/main/java/at/ac/tuwien/kr/alpha/grounder/structure/ProgramAnalysis.java]: I am not happy with the names here: `tmpInstances` are the interval-expanded instances/facts while the `tmpPredicate` is the predicate of all those facts.

File: src/main/java/at/ac/tuwien/kr/alpha/grounder/structure/ProgramAnalysis.java [hunk 10]
```diff
@@ -1,31 +1,81 @@
...
+			this.programFacts.putIfAbsent(tmpPredicate, new LinkedHashSet<>());
+			this.programFacts.get(tmpPredicate).addAll(tmpInstances);
+		}
+	}
+
+	private Map<Integer, NonGroundRule> initializeRules(Program program) {
+		Map<Integer, NonGroundRule> retVal = new HashMap<>();
+		NonGroundRule tmpRule;
```
Comment (AntoniusW) [inline src/main/java/at/ac/tuwien/kr/alpha/grounder/structure/ProgramAnalysis.java]: I think it is unnecessary to have the variable declaration outside of the loop (as regards performance I think the compiler can figure that out).

File: src/main/java/at/ac/tuwien/kr/alpha/grounder/structure/ProgramAnalysis.java [hunk 11]
```diff
@@ -1,31 +1,81 @@
...
+				this.recordDefiningRule(tmpRule.getHeadAtom().getPredicate(), tmpRule);
+			}
+		}
+		return retVal;
 	}
 
 	public void recordDefiningRule(Predicate headPredicate, NonGroundRule rule) {
-		predicateDefiningRules.putIfAbsent(headPredicate, new HashSet<>());
-		predicateDefiningRules.get(headPredicate).add(rule);
+		this.predicateDefiningRules.putIfAbsent(headPredicate, new HashSet<>());
```
Comment (AntoniusW) [inline src/main/java/at/ac/tuwien/kr/alpha/grounder/structure/ProgramAnalysis.java]: Is there a reason why all member accesses in this class are preceded by `this.` ?

File: src/main/java/at/ac/tuwien/kr/alpha/common/depgraph/DependencyGraph.java [hunk 12]
```diff
@@ -0,0 +1,202 @@
...
+	 * The transposed graph structure, i.e. the same set of nodes, but all edges reversed - needed for analysis of strongly connected components.
+	 */
+	private Map<Node, List<Edge>> transposedNodes;
+
+	private Map<Integer, List<Node>> stronglyConnectedComponents;
+
+	private int constraintNumber;
+
+	public DependencyGraph() {
+		// doing this since checkstyle doesn't let me initialize it directly to 0
```
Comment (AntoniusW) [inline src/main/java/at/ac/tuwien/kr/alpha/common/depgraph/DependencyGraph.java]: Checkstyle doesn't let you, because `0` is the default initialization value. If you put nothing it will be `0` anyways. This default constructor thus can be removed without changing anything.

File: src/main/java/at/ac/tuwien/kr/alpha/common/depgraph/DependencyGraph.java [hunk 13]
```diff
@@ -0,0 +1,202 @@
...
+				}
+				retVal.handleRuleBodyLiteral(tmpHeadNode, l);
+			}
+		}
+		retVal.findStronglyConnectedComponents();
+		return retVal;
+	}
+
+	public Node getNodeForPredicate(Predicate p) {
+		return this.nodesByPredicate.get(p);
```
Comment (AntoniusW) [inline src/main/java/at/ac/tuwien/kr/alpha/common/depgraph/DependencyGraph.java]: Is there a need for explicitly referring to `this` object?

File: src/main/java/at/ac/tuwien/kr/alpha/common/depgraph/DependencyGraph.java [hunk 14]
```diff
@@ -0,0 +1,202 @@
...
+	}
+
+	private Node handleRuleHead(NonGroundRule rule) {
+		LOGGER.trace("Processing head of rule: {}", rule);
+		Node retVal;
+		Predicate pred;
+		if (rule.isConstraint()) {
+			pred = this.generateConstraintDummyPredicate();
```
Comment (AntoniusW) [inline src/main/java/at/ac/tuwien/kr/alpha/common/depgraph/DependencyGraph.java]: There are many seemingly superfluous `this.` around, please remove them wherever possible.

File: src/main/java/at/ac/tuwien/kr/alpha/common/depgraph/DependencyGraph.java [hunk 15]
```diff
@@ -0,0 +1,202 @@
...
+		}
+		this.transposedNodes = transposed;
+	}
+
+	private Predicate generateConstraintDummyPredicate() {
+		Predicate retVal = Predicate.getInstance(String.format(DependencyGraph.CONSTRAINT_PREDICATE_FORMAT, this.nextConstraintNumber()), 0);
+		return retVal;
+	}
+
+	// TODO maybe move this to DependencyGraphUtils
```
Comment (AntoniusW) [inline src/main/java/at/ac/tuwien/kr/alpha/common/depgraph/DependencyGraph.java]: Please address this todo. Since this method relies a lot on `DependencyGraphUtils` I guess it could be part of that class too.

File: src/main/java/at/ac/tuwien/kr/alpha/common/depgraph/DependencyGraphUtils.java [hunk 16]
```diff
@@ -0,0 +1,147 @@
...
+import java.util.Deque;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+// TODO internalize into DependencyGraph - most stuff here depends on the dependency graph
```
Comment (AntoniusW) [inline src/main/java/at/ac/tuwien/kr/alpha/common/depgraph/DependencyGraphUtils.java]: Please address this todo: either move into `DependencyGraph` or keep it outside but give it a meaningful name, e.g. `DependencyGraphAlgorithms`

File: src/main/java/at/ac/tuwien/kr/alpha/common/depgraph/DependencyGraphUtils.java [hunk 17]
```diff
@@ -0,0 +1,147 @@
...
+	private static boolean isReachableFrom(Node dest, Node src, DependencyGraph dg, List<Node> discovered) {
+		List<Edge> outgoingEdges;
+		if (src.equals(dest)) {
+			return true;
+		}
+		if ((outgoingEdges = dg.getNodes().get(src)) == null) {
+			return false;
+		}
+		discovered.add(src);
+		// we wanna do BFS here, therefore use 2 loops
```
Comment (AntoniusW) [inline src/main/java/at/ac/tuwien/kr/alpha/common/depgraph/DependencyGraphUtils.java]: This could need some better explanation; also please use full sentences for comments.

File: src/main/java/at/ac/tuwien/kr/alpha/common/depgraph/DfsResult.java [hunk 18]
```diff
@@ -0,0 +1,28 @@
+package at.ac.tuwien.kr.alpha.common.depgraph;
+
+import java.util.Deque;
+import java.util.List;
+import java.util.Map;
+
```
Comment (AntoniusW) [inline src/main/java/at/ac/tuwien/kr/alpha/common/depgraph/DfsResult.java]: Please document the purpose of this class.

File: src/main/java/at/ac/tuwien/kr/alpha/common/depgraph/Edge.java [hunk 19]
```diff
@@ -0,0 +1,57 @@
+package at.ac.tuwien.kr.alpha.common.depgraph;
+
+/**
+ * An edge in a dependency graph.
+ */
+// TODO ensure immutability
```
Comment (AntoniusW) [inline src/main/java/at/ac/tuwien/kr/alpha/common/depgraph/Edge.java]: Please address this todo.

File: src/main/java/at/ac/tuwien/kr/alpha/common/depgraph/Edge.java [hunk 20]
```diff
@@ -0,0 +1,57 @@
...
+public class Edge {
+
+	private final Node target;
+	private final boolean sign;
+	private final String label;
```
Comment (AntoniusW) [inline src/main/java/at/ac/tuwien/kr/alpha/common/depgraph/Edge.java]: What is the label used for? It seems to be a `String` representation of the sign?

File: src/main/java/at/ac/tuwien/kr/alpha/common/depgraph/Node.java [hunk 21]
```diff
@@ -0,0 +1,134 @@
...
+/**
+ * A node in a dependency graph. One node references exactly one predicate. This means that all rule heads deriving the same predicate will be condensed into
+ * the same graph node. In some cases this results in more "conservative" results in stratification analysis, where some rules will not be evaluated up-front,
+ * although that would be possible.
+ * 
+ * Note that constraints are represented by one dummy predicate (named "constr_{num}"). Each constraint node has a negative edge to itself to express the
+ * notation of a constraint ":- a, b." as "x :- a, b, not x.".
+ *
+ */
+// TODO ensure immutability
```
Comment (AntoniusW) [inline src/main/java/at/ac/tuwien/kr/alpha/common/depgraph/Node.java]: Please address this todo.

File: src/main/java/at/ac/tuwien/kr/alpha/common/depgraph/Node.java [hunk 22]
```diff
@@ -0,0 +1,134 @@
...
+public class Node {
+
+	public static class NodeInfo {
+
+		private Node dfsPredecessor;
+		private int dfsDiscoveryTime;
+		private int dfsFinishTime;
+		private int componentId;
```
Comment (AntoniusW) [inline src/main/java/at/ac/tuwien/kr/alpha/common/depgraph/Node.java]: Except for the `componentId` no other member value seems to be accessed. Can this class be replaced with just an `int componentId`?

File: src/main/java/at/ac/tuwien/kr/alpha/common/depgraph/Node.java [hunk 23]
```diff
@@ -0,0 +1,134 @@
...
+		public NodeInfo() {
+
+		}
+
+		public NodeInfo(NodeInfo original) {
+			this.dfsDiscoveryTime = original.dfsDiscoveryTime;
+			this.dfsFinishTime = original.dfsFinishTime;
+		}
+
+		public int getDfsDiscoveryTime() {
```
Comment (AntoniusW) [inline src/main/java/at/ac/tuwien/kr/alpha/common/depgraph/Node.java]: Seems to be never called.

File: src/main/java/at/ac/tuwien/kr/alpha/common/depgraph/Node.java [hunk 24]
```diff
@@ -0,0 +1,134 @@
...
+			return this.dfsDiscoveryTime;
+		}
+
+		public void setDfsDiscoveryTime(int dfsDiscoveryTime) {
+			this.dfsDiscoveryTime = dfsDiscoveryTime;
+		}
+
+		public int getDfsFinishTime() {
```
Comment (AntoniusW) [inline src/main/java/at/ac/tuwien/kr/alpha/common/depgraph/Node.java]: Seems to be never called.

File: src/main/java/at/ac/tuwien/kr/alpha/common/depgraph/Node.java [hunk 25]
```diff
@@ -0,0 +1,134 @@
...
+			return this.dfsFinishTime;
+		}
+
+		public void setDfsFinishTime(int dfsFinishTime) {
+			this.dfsFinishTime = dfsFinishTime;
+		}
+
+		public Node getDfsPredecessor() {
```
Comment (AntoniusW) [inline src/main/java/at/ac/tuwien/kr/alpha/common/depgraph/Node.java]: Seems to be never called.

File: src/main/java/at/ac/tuwien/kr/alpha/common/depgraph/Node.java [hunk 26]
```diff
@@ -0,0 +1,134 @@
...
+	public Node(Predicate predicate) {
+		this(predicate, predicate.toString());
+	}
+
+	/**
+	 * Copy-constructor - constructs a new node as a deep-copy of the passed node
+	 * 
+	 * @param original the node to copy
+	 */
+	public Node(Node original) {
```
Comment (AntoniusW) [inline src/main/java/at/ac/tuwien/kr/alpha/common/depgraph/Node.java]: The copy constructor seems to be never used.

File: src/main/java/at/ac/tuwien/kr/alpha/common/depgraph/io/DependencyGraphWriter.java [hunk 27]
```diff
@@ -0,0 +1,86 @@
...
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.BiFunction;
+
+import at.ac.tuwien.kr.alpha.common.depgraph.DependencyGraph;
+import at.ac.tuwien.kr.alpha.common.depgraph.Edge;
+import at.ac.tuwien.kr.alpha.common.depgraph.Node;
+
+public class DependencyGraphWriter {
```
Comment (AntoniusW) [inline src/main/java/at/ac/tuwien/kr/alpha/common/depgraph/io/DependencyGraphWriter.java]: Please document this class.

File: src/test/java/at/ac/tuwien/kr/alpha/common/depgraph/ComponentGraphTest.java [hunk 28]
```diff
@@ -0,0 +1,265 @@
...
+		Assert.assertEquals(1, strata.size());
+		List<SCComponent> stratum0 = strata.get(0);
+		Assert.assertEquals(3, stratum0.size());
+	}
+
+	@Test
+	public void stratifyWithNegativeDependencyTest() throws IOException {
+		Alpha system = new Alpha();
+		StringBuilder bld = new StringBuilder();
+		bld.append("b :- a.").append("\n");
```
Comment (AntoniusW) [inline src/test/java/at/ac/tuwien/kr/alpha/common/depgraph/ComponentGraphTest.java]: You can actually use `+` for strings here, since the compiler will turn it into `StringBuilder` usage automatically (or even statically evaluate it and build one large `String` object) .

File: src/test/java/at/ac/tuwien/kr/alpha/common/depgraph/ComponentGraphTest.java [hunk 29]
```diff
@@ -0,0 +1,265 @@
...
+
+		Assert.assertEquals(1, strata.size());
+		List<SCComponent> stratum0 = strata.get(0);
+		assertNodesMatchStratumNodes(stratum0, ancestorOf, parentOf);
+	}
+
+	@Test
+	public void stratifyLargeGraphTest() {
+		Alpha system = new Alpha();
+		StringBuilder bld = new StringBuilder();
```
Comment (AntoniusW) [inline src/test/java/at/ac/tuwien/kr/alpha/common/depgraph/ComponentGraphTest.java]: Again, no need to do the string building manually.

File: src/test/java/at/ac/tuwien/kr/alpha/common/depgraph/DependencyGraphTest.java [hunk 30]
```diff
@@ -0,0 +1,310 @@
...
+				}
+			}
+			prgBuilder.append(".");
+			prgBuilder.append("\n");
+		}
+		return prgBuilder.toString();
+	}
+
+	@Test
+	@Ignore("Not a real test, rather a playground for local testing while changing stuff")
```
Comment (AntoniusW) [inline src/test/java/at/ac/tuwien/kr/alpha/common/depgraph/DependencyGraphTest.java]: This does not look ideal. Please put the private/local testing stuff into a file that is not committed to git.

File: src/test/java/at/ac/tuwien/kr/alpha/common/depgraph/DependencyGraphTest.java [hunk 31]
```diff
@@ -0,0 +1,310 @@
...
+		Program prog = system.readProgramString("b :- a.", null);
+		ProgramAnalysis pa = new ProgramAnalysis(prog);
+		DependencyGraph dg = pa.getDependencyGraph();
+
+		Node a = dg.getNodeForPredicate(Predicate.getInstance("a", 0));
+		Node b = dg.getNodeForPredicate(Predicate.getInstance("b", 0));
+
+		Node nonExistant = new Node(Predicate.getInstance("notHere", 0));
+
+		Assert.assertEquals(true, DependencyGraphUtils.isReachableFrom(a, a, dg));
```
Comment (AntoniusW) [inline src/test/java/at/ac/tuwien/kr/alpha/common/depgraph/DependencyGraphTest.java]: This may be simplified to `assertTrue` and `assertFalse`.

**Comment by madmike200590** [comment]
> Some `HeadBodyTransformationTests` are currently failing with the following exception:
> 
> ```
> java.lang.StackOverflowError
> 	at ch.qos.logback.classic.Logger.trace(Logger.java:429)
> 	at at.ac.tuwien.kr.alpha.solver.learning.GroundConflictNoGoodLearner.computeBackjumpingDecisionLevel(GroundConflictNoGoodLearner.java:294)
> 	at at.ac.tuwien.kr.alpha.solver.learning.GroundConflictNoGoodLearner.repeatAnalysisIfNotAssigning(GroundConflictNoGoodLearner.java:225)
> 	at at.ac.tuwien.kr.alpha.solver.learning.GroundConflictNoGoodLearner.analyzeConflictingNoGoodRepetition(GroundConflictNoGoodLearner.java:176)
> 	at at.ac.tuwien.kr.alpha.solver.learning.GroundConflictNoGoodLearner.repeatAnalysisIfNotAssigning(GroundConflictNoGoodLearner.java:229)
> 	at at.ac.tuwien.kr.alpha.solver.learning.GroundConflictNoGoodLearner.analyzeConflictingNoGoodRepetition(GroundConflictNoGoodLearner.java:176)
> 	at at.ac.tuwien.kr.alpha.solver.learning.GroundConflictNoGoodLearner.repeatAnalysisIfNotAssigning(GroundConflictNoGoodLearner.java:229)
> 	at at.ac.tuwien.kr.alpha.solver.learning.GroundConflictNoGoodLearner.analyzeConflictingNoGoodRepetition(GroundConflictNoGoodLearner.java:176)
> 	at at.ac.tuwien.kr.alpha.solver.learning.GroundConflictNoGoodLearner.repeatAnalysisIfNotAssigning(GroundConflictNoGoodLearner.java:229)
> ...
> ```
> Interestingly, the tests sometimes fail and sometimes succeed. Could this be a concurrency issue?

This is quite interesting - didn't occur locally for me and also doesn't look like anything I touched.
I`ll take a look at that (and all the other feedback on this PR during the second half of this week / early next week)
